rust/kernel/rbtree.rs

```diff
@@ -0,0 +1,540 @@
+    }
+
+    /// Inserts a new node into the tree.
+    ///
+    /// It overwrites a node if one already exists with the same key and returns it (containing the
+    /// key/value pair). Returns [`None`] if a node with the same key didn't already exist.
+    ///
+    /// This function always succeeds.
+    pub fn insert(&mut self, node: RBTreeNode<K, V>) -> Option<RBTreeNode<K, V>>
+    where
+        K: Ord,
+    {
+        let RBTreeNode { node } = node;
+        // SAFETY: The pointer returned by `Box::into_raw` is valid at least until we call
+        // `Box::from_raw`, which only happens when the node is removed or replaced.
+        let node = unsafe { &mut *Box::into_raw(node) };
+        let mut new_link: &mut *mut bindings::rb_node = &mut self.root.rb_node;
+        let mut parent = core::ptr::null_mut();
+        while !new_link.is_null() {
+            let this = crate::container_of!(*new_link, Node<K, V>, links);
+
+            parent = *new_link;
+
+            // SAFETY: `this` is a non-null node so it is valid by the type invariants.
+            match node.key.cmp(unsafe { &(*this).key }) {
+                // SAFETY: `parent` is a non-null node so it is valid by the type invariants.
+                Ordering::Less => new_link = unsafe { &mut (*parent).rb_left },
+                // SAFETY: `parent` is a non-null node so it is valid by the type invariants.
+                Ordering::Greater => new_link = unsafe { &mut (*parent).rb_right },
+                Ordering::Equal => {
+                    // INVARIANT: We are replacing an existing node with a new one, which is valid.
+                    // It remains valid because we "forgot" it with `Box::into_raw`.
+                    // SAFETY: All pointers are non-null and valid (parent, despite the name, really
+                    // is the node we're replacing).
+                    unsafe { bindings::rb_replace_node(parent, &mut node.links, &mut self.root) };
+
+                    // INVARIANT: The node is being returned and the caller may free it, however,
+                    // it was removed from the tree. So the invariants still hold.
+                    return Some(RBTreeNode {
+                        // SAFETY: `this` was a node in the tree, so it is valid.
+                        node: unsafe { Box::from_raw(this as _) },
+                    });
+                }
+            }
+        }
+
+        // INVARIANT: We are linking in a new node, which is valid. It remains valid because we
+        // "forgot" it with `Box::into_raw`.
+        // SAFETY: All pointers are non-null and valid (`*new_link` is null, but `new_link` is a
+        // mutable reference).
+        unsafe { rust_helper_rb_link_node(&mut node.links, parent, new_link) };
+
+        // SAFETY: All pointers are valid. `node` has just been inserted into the tree.
+        unsafe { bindings::rb_insert_color(&mut node.links, &mut self.root) };
```

Comment:
The mutable reference to `node.links` is invalidated after this function call, but the function stores it in `self.root`, right? I think you should use `core::ptr::addr_of_mut!(node.links)` instead. Note that the `let node = unsafe { &mut *Box::into_raw(node) };` is also invalidated once this function returns. So maybe `let node = Box::into_raw(node);` and then `addr_of_mut!((*node).links)`? The `&mut self.root` is fine if there are no pointers to `self.root` only pointers stored in `self.root`.

Reply:
I think I need to update my mental model here.

What is invalidating `node.links`? How would using `addr_of_mut!(node.links)` avoid this invalidation?

Reply:
Thinking a bit more about this, this `&mut node.links` doesn't get invalidated, but the previous `&mut node.links` does due to this `&mut node.links`. `addr_of_mut!(node.links)` avoids this problem by directly creating a raw pointer without an intermediate mutable reference. It is valid to have multiple raw pointers to the same value, but creating one mutable reference invalidates all other references and pointers.

Reply:
What model are you using to determine what gets invalidated?

I quickly read the model in the [stacked borrows](https://plv.mpi-sws.org/rustbelt/stacked-borrows/paper.pdf) paper and my understanding is that the code is ok now. The reason is that the borrow stack for each byte that makes up `node` contains: `[..., SharedRW, Unique(n), ...]`, where `SharedRW` refers to the raw pointer returned by `into_raw()` and `Unique(n)` refers to the mutable reference created on line 248. So any new raw references that I create in the function can still be used after the function returns because they will be resolved by the `SharedRW` entry (until I call `from_raw()`, which will pop that `SharedRW` entry) because raw pointers are not "tagged".

IOW, I don't agree with your claim that "creating one mutable reference invalidates all other pointers" -- my understanding is that it invalidates all other pointers on top of it in the borrow stack. Pointers below it continue to be valid until something else pops them off.

I am new to stacked borrows though, so my understanding may be wrong. Where does it not match yours?

Reply:
> I quickly read the model in the stacked borrows paper and my understanding is that the code is ok now. The reason is that the borrow stack for each byte that makes up node contains: [..., SharedRW, Unique(n), ...], where SharedRW refers to the raw pointer returned by into_raw() and Unique(n) refers to the mutable reference created on line 248. So any new raw references that I create in the function can still be used after the function returns because they will be resolved by the SharedRW entry (until I call from_raw(), which will pop that SharedRW entry) because raw pointers are not "tagged".

You will end up with `SharedRW(k), Unique(l), SharedRW(m)` in this case. It  is fine with the default options of miri as the raw pointer can resolve to either `SharedRW`, but it isn't when `-Zmiri-track-raw-pointers` as it will only resolve to `SharedRW(m)` which is popped of the borrow stack at the next `&mut item.links` in addition `SharedRW(m)` only covers `links` and not the whole item, so accessing the rest of the item through it is not allowed.

> IOW, I don't agree with your claim that "creating one mutable reference invalidates all other pointers" -- my understanding is that it invalidates all other pointers on top of it in the borrow stack. Pointers below it continue to be valid until something else pops them off.

Yeah, I was simplifying a bit. Maybe a bit too much.

Reply:
> You will end up with `SharedRW(k), Unique(l), SharedRW(m)` in this case. It is fine with the default options of miri as the raw pointer can resolve to either `SharedRW`, but it isn't when `-Zmiri-track-raw-pointers` as it will only resolve to `SharedRW(m)` which is popped of the borrow stack at the next `&mut item.links` in addition `SharedRW(m)` only covers `links` and not the whole item, so accessing the rest of the item through it is not allowed.

From miri's [readme](https://github.com/rust-lang/miri/blob/master/README.md) about `-Zmiri-track-raw-pointers`: _This can make valid code fail to pass the checks_. So I'm not so concerned about it, I'm concerned about the code correctness wrt to the model. Do we agree that it is correct (wrt the model)?

Reply:
Part of the reason I declared a `&mut Node` was to avoid having to add more `SAFETY` justifications in each of its uses (4 other places), but now I realise that 3 of them are `&mut node.links` so I can avoid repeating the justification by storing a raw pointer in a variable and reusing it.

I'll make this change and upload it shortly.

Reply:
The new version is UB free I think.